### PR TITLE
[Enhancement] Move cte inline to memo phase 1

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Move cte inline rule to memo phase, we will enforce all inline or all reuse cte

1. add CTENoOP implements
2. add CTEProperty in Enforce phase
3. add CTE costs model

Opimizer Performance:

|  Query  | Before | After | 
|  ----  | ----  | ---- |
|  multi-nest view query with cte(8+) #9347  | 300ms | 300ms |
| 20 iterate cte + 60 refs  | 100ms | - |
| 12 iterate cte + 36 refs  | - | 2224ms  |
| 11 iterate cte + 33 refs  | - | 700ms  |
| 10 iterate cte + 30 refs  | - | 280ms  |
| 8 nested cte + 3 * 2^8 refs  | 4081ms | 7522ms  |
| 7 nested cte + 3 * 2^7 refs  | 1481ms | 2250ms  |
| 6 nested cte + 3 * 2^6 refs  | 451ms | 600ms  |

Of course, the cost of these test is in RuleWrite phase, not in memo phase.It's will take more time if we test so many joins or complex SQL

BTW, we need optimize the cost of memo merge 

SQLs
```
# iterater cte
with x1 as (select * from t0),
     x2 as (select * from t0),
     x3 as (select * from t0),
     x4 as (select * from t0),
     x5 as (select * from t0),
     x6 as (select * from t0),                    
     x7 as (select * from t0),
     x8 as (select * from t0),          
     x9 as (select * from t0),          
     x10 as (select * from t0),
     x11 as (select * from t0)     
select * from x1 union all
select * from x1 union all
select * from x1 union all
select * from x2 union all
select * from x2 union all
select * from x2 union all
select * from x3 union all
select * from x3 union all
select * from x3 union all
select * from x4 union all
select * from x4 union all
select * from x4 union all
select * from x5 union all
select * from x5 union all
select * from x5 union all
select * from x6 union all
select * from x6 union all
select * from x6 union all
select * from x7 union all
select * from x7 union all
select * from x7 union all
select * from x8 union all
select * from x8 union all
select * from x8 union all
select * from x9 union all
select * from x9 union all
select * from x9 union all
select * from x10 union all
select * from x10 union all
select * from x10 union all
select * from x11 union all
select * from x11 union all
select * from x11;

# nest cte
explain 
with x1 as (select * from t0),
     x2 as (select * from x1 union all 
            select * from x1),
     x3 as (select * from x2 union all 
            select * from x2),
     x4 as (select * from x3 union all 
            select * from x3),
     x5 as (select * from x4 union all 
            select * from x4),
     x6 as (select * from x5 union all 
            select * from x5),
     x7 as (select * from x6 union all 
            select * from x6)
select * from x1 union all
select * from x1 union all
select * from x1 union all
select * from x2 union all
select * from x2 union all
select * from x2 union all
select * from x3 union all
select * from x3 union all
select * from x3 union all
select * from x4 union all
select * from x4 union all
select * from x4 union all
select * from x5 union all
select * from x5 union all
select * from x5 union all
select * from x6 union all
select * from x6 union all
select * from x6 union all
select * from x7 union all
select * from x7 union all
select * from x7 ;


```

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 174 / 189 (92.06%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEConsumeOperator.java|0|1|00.00%|[84]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineCTEConsumeRule.java|0|4|00.00%|[14, 17, 21, 22]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalCTEAnchorOperator.java|1|2|50.00%|[29]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalNoCTEOperator.java|4|5|80.00%|[43]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|24|29|82.76%|[53, 58, 59, 68, 81]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|12|13|92.31%|[1364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|14|15|93.33%|[215]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/PhysicalPropertySet.java|14|15|93.33%|[97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumerReuseImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorToNoCTEImplementationRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumeInlineImplementationRule.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/SeriallyTaskScheduler.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineOneCTEConsumeRule.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEContext.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEUtils.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 173 / 188 (92.02%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEConsumeOperator.java|0|1|00.00%|[84]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineCTEConsumeRule.java|0|4|00.00%|[14, 17, 21, 22]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalCTEAnchorOperator.java|1|2|50.00%|[29]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalNoCTEOperator.java|4|5|80.00%|[43]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|24|29|82.76%|[53, 58, 59, 68, 81]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|12|13|92.31%|[1364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|14|15|93.33%|[215]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/PhysicalPropertySet.java|14|15|93.33%|[97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumerReuseImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorToNoCTEImplementationRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumeInlineImplementationRule.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineOneCTEConsumeRule.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEContext.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEUtils.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    public Set<Integer> getCteIds() {
+        return cteIds;
+    }
+
+    public CTEProperty removeCTE(int cteID) {
+        CTEProperty p = new CTEProperty();
+        p.getCteIds().addAll(this.cteIds);
+        p.getCteIds().removeIf(c -> c.equals(cteID));
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java:39]: Why not directly remove the cteID?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        return cteIds.isEmpty();
+    }
+
+    public void merge(CTEProperty other) {
+        this.cteIds.addAll(other.cteIds);
+    }
+
+    @Override
+    public boolean isSatisfy(PhysicalProperty other) {
+        return true;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java:53]: why always true?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java [hunk 3]
```diff
@@ -258,4 +263,48 @@ public Void visitPhysicalNoCTE(PhysicalNoCTEOperator node, ExpressionContext con
...
+        public Void visitPhysicalCTEAnchor(PhysicalCTEAnchorOperator node, Void context) {
+            visitOperator(node, context);
+            PhysicalPropertySet requiredRight = requiredProperties.get(0).get(1);
+
+            CTEProperty thisCTE = new CTEProperty(node.getCteId());
+            thisCTE.merge(requiredRight.getCteProperty());
+            PhysicalPropertySet copy = requiredRight.copy();
+
+            copy.setCteProperty(thisCTE);
+            requiredProperties.get(0).set(1, copy);
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java:296]: If you set the copy property back, why do we need the copy?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+import com.starrocks.sql.optimizer.Group;
+import com.starrocks.sql.optimizer.GroupExpression;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class CTEProperty implements PhysicalProperty {
+    private final Set<Integer> cteIds;
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java:16]: comment why CTEProperty only need the cteIds.

**Comment by kangkaisen** [comment]
Need the optimizer itself performance test, we need to know the optimizer will slow how much time.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java (hunk 3 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java:296]: Don't modify origin property(maybe global property: EMPTY)

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java (hunk 2 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java:53]: Never call the method, and CTE property will prune invalid plan directly, don't neet enforce

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/base/CTEProperty.java:39]: Don't modify self

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 173 / 188 (92.02%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEConsumeOperator.java|0|1|00.00%|[84]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineCTEConsumeRule.java|0|4|00.00%|[14, 17, 21, 22]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalCTEAnchorOperator.java|1|2|50.00%|[29]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalNoCTEOperator.java|4|5|80.00%|[43]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|24|29|82.76%|[54, 59, 60, 69, 82]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|12|13|92.31%|[1364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|14|15|93.33%|[215]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/PhysicalPropertySet.java|14|15|93.33%|[97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumerReuseImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorToNoCTEImplementationRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumeInlineImplementationRule.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|11|11|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineOneCTEConsumeRule.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEContext.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEUtils.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|4|4|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java [hunk 5]
```diff
@@ -329,14 +315,15 @@ public CostEstimate visitPhysicalCTEProduce(PhysicalCTEProduceOperator node, Exp
 
         @Override
         public CostEstimate visitPhysicalCTEAnchor(PhysicalCTEAnchorOperator node, ExpressionContext context) {
-            return CostEstimate.zero();
+            // memory cost
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:318]: This cost why not compute at CTE producer?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java [hunk 6]
```diff
@@ -211,9 +203,12 @@ private void enforceChildDistribution(DistributionSpec distributionSpec, GroupEx
             }
         } else {
             // add physical distribution operator
-            addChildEnforcer(childOutputProperty, new DistributionProperty(distributionSpec),
-                    childCosts, child.getGroup());
+            GroupExpression enforcer = newDistributionProperty.appendEnforcers(childGroup);
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java:206]: `addChildEnforcer` the function will remove next pr

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java [hunk 7]
```diff
@@ -135,6 +135,7 @@ void memoOptimize(ConnectContext connectContext, Memo memo, TaskContext rootTask
         if (!sessionVariable.isDisableJoinReorder()
                 && Utils.countInnerJoinNodeSize(tree) < sessionVariable.getCboMaxReorderNode()) {
             if (Utils.countInnerJoinNodeSize(tree) > sessionVariable.getCboMaxReorderNodeUseExhaustive()) {
+                CTEUtils.collectCteStatistics(memo, context);
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java:138]: join reorder need statistics

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/CTEConsumeInlineImplementationRule.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalCTEConsumeOperator consume = (LogicalCTEConsumeOperator) input.getOp();
+        return !context.getCteContext().isForceCTE(consume.getCteId());
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        // to NoOP
+        LogicalCTEConsumeOperator logical = (LogicalCTEConsumeOperator) input.getOp();
+        PhysicalNoCTEOperator noCTEOperator = new PhysicalNoCTEOperator(logical.getCteId(), logical.getProjection());
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/CTEConsumeInlineImplementationRule.java:32]: to inline cte plan, and mark this plan can't satisfy cte property

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java [hunk 9]
```diff
@@ -258,4 +263,48 @@ public Void visitPhysicalNoCTE(PhysicalNoCTEOperator node, ExpressionContext con
...
+            return null;
+        }
+
+        @Override
+        public Void visitPhysicalCTEAnchor(PhysicalCTEAnchorOperator node, Void context) {
+            visitOperator(node, context);
+            PhysicalPropertySet requiredRight = requiredProperties.get(0).get(1);
+
+            CTEProperty thisCTE = new CTEProperty(node.getCteId());
+            thisCTE.merge(requiredRight.getCteProperty());
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java:292]: yeah, all cteid need pass to children

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/RequiredPropertyDeriver.java:296]: Anchor don't need required produce

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java (hunk 5 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/cost/CostModel.java:318]: of course if you want

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 189 / 207 (91.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEConsumeOperator.java|0|1|00.00%|[84]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineCTEConsumeRule.java|0|4|00.00%|[14, 17, 21, 22]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalCTEAnchorOperator.java|1|2|50.00%|[29]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalNoCTEOperator.java|4|5|80.00%|[43]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|24|29|82.76%|[54, 59, 60, 69, 82]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|27|30|90.00%|[148, 149, 150]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|12|13|92.31%|[1364]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|14|15|93.33%|[215]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/PhysicalPropertySet.java|14|15|93.33%|[97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumerReuseImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorToNoCTEImplementationRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumeInlineImplementationRule.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineOneCTEConsumeRule.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEContext.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEUtils.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|4|4|100.00%|[]|

**Review by Youngwb — DISMISSED** [review]

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 189 / 207 (91.30%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEConsumeOperator.java|0|1|00.00%|[84]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineCTEConsumeRule.java|0|4|00.00%|[14, 17, 21, 22]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalCTEAnchorOperator.java|1|2|50.00%|[29]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalNoCTEOperator.java|4|5|80.00%|[43]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|24|29|82.76%|[54, 59, 60, 69, 82]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|27|30|90.00%|[148, 149, 150]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|12|13|92.31%|[1372]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|14|15|93.33%|[215]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/PhysicalPropertySet.java|14|15|93.33%|[97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|26|26|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumerReuseImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorToNoCTEImplementationRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumeInlineImplementationRule.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineOneCTEConsumeRule.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEContext.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEUtils.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|4|4|100.00%|[]|

**Review by liuyehcf — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java [hunk 10]
```diff
@@ -182,9 +193,7 @@ public boolean needPushLimit() {
      * Inline CTE consume sense:
      * 1. Disable CTE reuse, must inline all CTE
      * 2. CTE consume only use once, it's meanings none CTE data reuse
-     * 3. CTE produce cost > sum of all consume inline costs, should considered CTE extract cost(network/memory),
-     *    so produce cost will multi a rate
-     *
+     * 3. CTE ratio is zero
```
[the remark this answers is not in the document]
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEContext.java:196]: use to adjust the costs of CTE

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by Seaven** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 190 / 208 (91.35%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/logical/LogicalCTEConsumeOperator.java|0|1|00.00%|[84]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineCTEConsumeRule.java|0|4|00.00%|[14, 17, 21, 22]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalCTEAnchorOperator.java|1|2|50.00%|[29]|
|:large_blue_circle:|com/starrocks/sql/optimizer/operator/physical/PhysicalNoCTEOperator.java|4|5|80.00%|[43]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/CTEProperty.java|24|29|82.76%|[54, 59, 60, 69, 82]|
|:large_blue_circle:|com/starrocks/sql/optimizer/ChildOutputPropertyGuarantor.java|27|30|90.00%|[148, 149, 150]|
|:large_blue_circle:|com/starrocks/sql/optimizer/statistics/StatisticsCalculator.java|12|13|92.31%|[1372]|
|:large_blue_circle:|com/starrocks/sql/optimizer/task/EnforceAndCostTask.java|14|15|93.33%|[215]|
|:large_blue_circle:|com/starrocks/sql/optimizer/base/PhysicalPropertySet.java|14|15|93.33%|[97]|
|:large_blue_circle:|com/starrocks/sql/optimizer/RequiredPropertyDeriver.java|27|27|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|5|5|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumerReuseImplementationRule.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorToNoCTEImplementationRule.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEConsumeInlineImplementationRule.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/implementation/CTEAnchorImplementationRule.java|3|3|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/OutputPropertyDeriver.java|24|24|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/InlineOneCTEConsumeRule.java|6|6|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEContext.java|8|8|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/CTEUtils.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/cost/CostModel.java|4|4|100.00%|[]|

**Comment by Seaven** [comment]
run starrocks_admit_test

**Pull request merged**